src/platform/FailSafeContext.cpp

```diff
@@ -37,36 +37,32 @@ constexpr TLV::Tag kUpdateNocCommandTag = TLV::ContextTag(2);
 
 void FailSafeContext::HandleArmFailSafeTimer(System::Layer * layer, void * aAppState)
 {
-    FailSafeContext * context = reinterpret_cast<FailSafeContext *>(aAppState);
-    context->FailSafeTimerExpired();
+    FailSafeContext * failSafeContext = reinterpret_cast<FailSafeContext *>(aAppState);
+    failSafeContext->FailSafeTimerExpired();
 }
 
 void FailSafeContext::HandleDisarmFailSafe(intptr_t arg)
 {
-    FailSafeContext * this_ = reinterpret_cast<FailSafeContext *>(arg);
-
-    this_->mFailSafeBusy = false;
-
-    if (ConfigurationMgr().SetFailSafeArmed(false) != CHIP_NO_ERROR)
-    {
-        ChipLogError(DeviceLayer, "Failed to set FailSafeArmed config to false");
-    }
-
-    if (DeleteFromStorage() != CHIP_NO_ERROR)
-    {
-        ChipLogError(DeviceLayer, "Failed to delete FailSafeContext from configuration");
-    }
+    FailSafeContext * failSafeContext = reinterpret_cast<FailSafeContext *>(arg);
+    failSafeContext->DisarmFailSafe();
 }
 
 void FailSafeContext::FailSafeTimerExpired()
 {
+    if (!IsFailSafeArmed())
+    {
+        // In case this was a pending timer event in event loop, and we had
+        // done CommissioningComplete or manual disarm.
+        return;
+    }
+
+    ChipLogProgress(FailSafe, "Fail-safe timer expired");
+    DeviceLayer::SystemLayer().CancelTimer(HandleArmFailSafeTimer, this);
     ScheduleFailSafeCleanup(mFabricIndex, mAddNocCommandHasBeenInvoked, mUpdateNocCommandHasBeenInvoked);
 }
 
 void FailSafeContext::ScheduleFailSafeCleanup(FabricIndex fabricIndex, bool addNocCommandInvoked, bool updateNocCommandInvoked)
 {
-    ResetState();
```

Comment:
Why is this being removed?  That will make it look like a fail-safe is still armed, when it's critically important at this point that it _not_ look like it's armed.

Reply:
That's why there is the "busy" flag below.

 The ResetState is done by     `PlatformMgr().ScheduleWork(HandleDisarmFailSafe, reinterpret_cast<intptr_t>(this));` --> HandleDisarmFailSafe --> DisarmFailSafe via the event loop.

I hate this and we should fix it... But not this PR

Reply:
The busy flag is fine, but the point of this ResetState was to clear the "armed" state.  We want `IsFailSafeArmed()` to test false starting at this point.  That's why the ResetState was there.

That means either setting `mFailSafeArmed` to false or having `IsFailSafeArmed()` test the busy flag or something....

Reply:
Fixed

Reply:
Fixed and added a bit of docs. The solution is to not post anything to even loop anymore...